ably/options.go

```diff
@@ -14,36 +15,86 @@ const (
 	ProtocolMsgPack = "msgpack"
 )
 
+var errInvalidKey = errors.New("invalid key format")
+
+var DefaultOptions = &ClientOptions{
+	RestEndpoint:     "rest.ably.io",
+	RealtimeEndpoint: "realtime.ably.io",
+	Protocol:         ProtocolJSON, // TODO: make it ProtocolMsgPack
+}
+
 type ClientOptions struct {
-	RealtimeEndpoint string
 	RestEndpoint     string
-	Key              string
-	Token            string
-	Secret           string
+	RealtimeEndpoint string
+	KeyName          string
+	KeySecret        string
 	ClientID         string
-	Protocol         string
+	Protocol         string // either ProtocolJSON or ProtocolMsgPack
 	UseTokenAuth     bool
 	NoTLS            bool
 
 	HTTPClient *http.Client
 }
 
-func (p *ClientOptions) Prepare() {
-	if p.Key != "" {
-		p.parseKey()
+func NewClientOptions(key string) (*ClientOptions, error) {
+	opts := &ClientOptions{}
+	if err := opts.SetKey(key); err != nil {
+		return nil, err
+	}
+	return opts, nil
+}
+
+func (opts *ClientOptions) restURL() string {
+	host := opts.RestEndpoint
+	if host == "" {
+		host = DefaultOptions.RestEndpoint
+	}
+	if opts.NoTLS {
+		return "http://" + host
+	}
+	return "https://" + host
+}
+
+func (opts *ClientOptions) realtimeURL() string {
+	host := opts.RealtimeEndpoint
+	if host == "" {
+		host = DefaultOptions.RealtimeEndpoint
+	}
+	if opts.NoTLS {
+		return "ws://" + host + ":80"
+	}
+	return "wss://" + host + ":443"
+}
+
+func (opts *ClientOptions) SetKey(key string) error {
```

Comment:
But still RestClient uses its own Options field to e.g. read RestEndpoint. If user changes RestClient.Options.RestEndpoint then there's no way to prevent that other than documenting it that it must be accessed only for reading:

```go
options := ably.NewClientOptions()
options.Key = "something"

client, err := ably.NewRestClient(options)

client.Options.RestEndpoint = "invalid" // this is going to break all the requests made by this client
```